packages/clay-dropdown/src/ClayDropdownBase.js

```diff
@@ -123,6 +123,180 @@ class ClayDropdownBase extends ClayComponent {
+
+				subActive = 0;
+			}
+		}
+
+		return {active, subActive};
+	}
+
+	/**
+	 * Gets the indexes of the next active item.
+	 * @param {Array} items
+	 * @param {Number} initialActive
+	 * @param {Number} initialSubActive
+	 * @param {Boolean} reverse
+	 * @private
+	 * @return {Number}
+	 */
+	_getNextIndexes(items, initialActive, initialSubActive, reverse = false) {
+		let {active, subActive} = this._incrementActiveItemIndex(
+			items,
+			initialActive === -1 ? 0 : initialActive,
+			initialSubActive,
+			reverse
+		);
+
+		const totalItems = items.flat(2);
+
+		let tick = 0;
+
+		while (items[active][subActive].disabled && tick < totalItems.length) {
+			const item = this._incrementActiveItemIndex(
+				items,
+				active,
+				subActive,
+				reverse
+			);
+
+			active = item.active;
+			subActive = item.subActive;
+
+			tick++;
+		}
+
+		return {active, subActive};
+	}
+
+	/**
+	 * Sets the next item in the list as active.
+	 * @param {Boolean} reverse
+	 * @private
+	 */
+	_setNextActive(reverse) {
+		const items = this._flattenItems();
+		let activeSubIndex = -1;
+
+		const activeIndex = items.findIndex(item => {
+			const subIndex = item.findIndex(item => item.active);
+
+			if (subIndex !== -1) {
+				activeSubIndex = subIndex;
+
+				return true;
+			}
+		});
+
+		const {active, subActive} = this._getNextIndexes(
+			items,
+			activeIndex,
+			activeSubIndex,
+			reverse
+		);
+
+		if (items[activeIndex] && items[activeIndex][activeSubIndex]) {
+			items[activeIndex][activeSubIndex].active = false;
+		}
+
+		items[active][subActive].active = true;
+
+		this.items = this._assembleFromFlattenedItems(items);
+	}
+
+	/**
+	 * Flattens all items into an array of arrays.
+	 * @private
+	 * @return {Array}
+	 */
+	_flattenItems() {
+		return this.items.map(item => (item.items ? item.items : [item]));
+	}
+
+	/**
+	 * Reassembles flattened items into the correct items structure.
+	 * @param {Array} flattenedItems
+	 * @private
+	 * @return {Array}
+	 */
+	_assembleFromFlattenedItems(flattenedItems) {
+		return this.items.map((item, i) => {
+			if (item.items) {
+				item.items = flattenedItems[i];
+			} else {
+				item = flattenedItems[i][0];
+			}
+
+			return item;
+		});
+	}
+
+	/**
+	 * Handle key events.
+	 * @param {!Event} event
+	 * @private
+	 */
+	_handleKeyNavigation(event) {
+		event.preventDefault();
+
+		if (event.key === 'ArrowDown' || event.key === 'ArrowUp') {
+			this._setNextActive(event.key === 'ArrowUp');
+		} else if (event.key === 'Enter') {
+			this._handleCheckedToggle();
+		}
+	}
+
+	/**
+	 * Toggles the 'checked' attribute on the active item.
+	 * @private
+	 */
+	_handleCheckedToggle() {
+		const flattenedItems = this._flattenItems();
+
+		for (let i = 0; i < flattenedItems.length; i++) {
+			const data = flattenedItems[i];
+
+			for (let j = 0; j < data.length; j++) {
+				const item = data[j];
+
+				if (item.active) {
+					item.checked = !item.checked;
```

Comment:
Hmm, not sure what you mean here by  "once it is used with incremental components"?

Shouldn't the DOM reflect the state of our component? Otherwise we will be managing state in two different places, both the in the DOM and in the state of the component. Our state already manages `active`, `disabled`, and `inputValue` of each item.

I may not be totally familiar with our patterns for storing state though. I am used to a more reactive way where the DOM is a reflection of our state.

Reply:
Maybe I was not so clear with that 😅, but when I say incremental components, they are components of Metal.

You're right about this... But in the case of checked is that we're manipulating the state with an action inside our component. This way the state can be lost when a new rendering occurs in the application and thus the item can lose the checked. The Parent component that is calling ClayDropdown will be in a different state from what we are manipulating in the items that are checked, for example. Let me know if I'm being clear on this.

When the item is clicked (_handleItemClick) we propagate this action so that the developer decides what to do with this information, we should do the same when we manipulate item checked so that the developer can do some action and update its state and also to keep the consistency of events.

Reply:
Oh! I see what you mean. That totally makes sense. I should've looked at `_handleItemClick ` more in depth to see that it was propagating to the parent.

Thanks for the clarification!